### PR TITLE
High-order Nodal Fields and Solution Transfer

### PR DESCRIPTION
**Pull request by mortezah**
This pull request achieves the following

- adds higher order nodal (hierarchic) fields similar to the implementation of MFEM
- adds a regression test for such fields
- updates the solution transfer codes so that these higher order fields are also transferred during adapt 
- adds a regression test for higher order solutions transfer

### CODE REVIEW

**Comment by cwsmith** [comment]
Looks good at first glance.  I'll take a closer look tomorrow. 

Please add any new source files to the `pkg_tribits.cmake` files in each directory.

**Comment by mortezah** [comment]
@cwsmith the new source are in the pkg_tribits.cmake

**Comment by mortezah** [comment]
@cwsmith 

Do you know when this would be merged? A student I am working with will need these changes for his development.

**Comment by cwsmith** [comment]
Sorry for the delay.  I'll look in about an hour.

File: apf/apfH1Shapes.cc [hunk 1]
```diff
@@ -0,0 +1,769 @@
...
+    getChebyshevT(p, x, &shape_x[0]);
+    getChebyshevT(p, y, &shape_y[0]);
+    getChebyshevT(p, z, &shape_z[0]);
+    getChebyshevT(p, 1. - x - y - z, &shape_l[0]);
+
+    for (int k = 0; k <= p; k++)
+      for (int j = 0; j + k <= p; j++)
+        for (int i = 0; i + j + k <= p; i++)
+          T(o++, m) = shape_x[i]*shape_y[j]*shape_z[k]*shape_l[p-i-j-k];
+	}
```
Comment (cwsmith) [inline apf/apfH1Shapes.cc:267]: Is there a tab here?  The indentation looks odd.

File: apf/apfH1Shapes.cc [hunk 2]
```diff
@@ -0,0 +1,769 @@
...
+#include "apfVectorElement.h"
+#include "apfPolyBasis1D.h"
+#include <mth.h>
+#include <mth_def.h>
+#include <mthQR.h>
+#include <pcu_util.h>
+#include <PCU.h>
+
+#include <iostream>
+using namespace std;
```
Comment (cwsmith) [inline apf/apfH1Shapes.cc:21]: Is this needed?  I don't see much stl usage below.

File: apf/apfH1Shapes.cc [hunk 3]
```diff
@@ -0,0 +1,769 @@
...
+
+  apf::NewArray<double> cp;
+  getClosedPoints(P, cp);
+
+  if (type == apf::Mesh::EDGE) {
+    PCU_ALWAYS_ASSERT(node >= 0 && node < countInternalEdgeNodes(P));
+    int c = 0;
+    for (int i = 1; i < P; i++)
+      if (node == c)
+      	return apf::Vector3(2*cp[i]-1, 0., 0.);
```
Comment (cwsmith) [inline apf/apfH1Shapes.cc:327]: tabs?  github does not always render the text white space correctly

File: apf/apfH1Shapes.cc [hunk 4]
```diff
@@ -0,0 +1,769 @@
...
+      }
+    };
+    class Edge : public apf::EntityShape
+    {
+    public:
+      int getOrder() {return P;}
+      void getValues(apf::Mesh* /*m*/, apf::MeshEntity* /*e*/,
+	  apf::Vector3 const& xi, apf::NewArray<double>& shapes) const
+      {
+      	const int p = P;
```
Comment (cwsmith) [inline apf/apfH1Shapes.cc:406]: tabs?

File: apf/apfH1Shapes.cc [hunk 5]
```diff
@@ -0,0 +1,769 @@
...
+        for (int j = 0; j <= p; j++)
+          for (int i = 0; i + j <= p; i++)
+            u(n++, 0) = shape_x[i]*shape_y[j]*shape_l[p-i-j];
+
+	      mth::Matrix<double> Q(dof, dof);
+      	mth::Matrix<double> R(dof, dof);
+      	getTi(P, apf::Mesh::TRIANGLE, Q, R);
+
+        mth::Matrix<double> S(dof, 1);
+      	for(int i = 0; i < 1; i++) // S = Ti * u
```
Comment (cwsmith) [inline apf/apfH1Shapes.cc:465]: tabs?

File: ma/maSolutionTransfer.cc [hunk 6]
```diff
@@ -131,7 +131,11 @@ class CavityTransfer : public FieldTransfer
       int bestI = 0;
       for (int i = 0; i < n; ++i)
       {
-        Vector xi = elemInvMaps[i] * point;
+        Vector xi;
+        if (mesh->getShape()->getOrder() == 1)
```
Comment (cwsmith) [inline ma/maSolutionTransfer.cc:135]: the indentation looks odd here - it could be github

File: test/H1Shapes.cc [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+#include <lionPrint.h>
+#include <pcu_util.h>
+
+#include <cassert>
+#include <cstdlib>
+#include <iostream>
+#include <string>
+#include <math.h>
+
+using namespace std;
```
Comment (cwsmith) [inline test/H1Shapes.cc:18]: is this needed?

File: test/testing.cmake [hunk 8]
```diff
@@ -526,6 +526,14 @@ mpi_test(l2_shape_tet_parallel 4
   ./L2Shapes
   ".null"
   "${MDIR}/4/cube.smb")
+mpi_test(h1_shape_serial 1
+  ./H1Shapes
+  ".null"
```
Comment (cwsmith) [inline test/testing.cmake:531]: any reason not to load the model? (same for the following test)

**Review by cwsmith — APPROVED** [review]
I didn't see any obvious major issues.  

There are a few spots to check for indentation.  

The use of `using namespace std` at the top of a source (and definitely a header) is generally not great since it 'infects' the entire file with the contents of `std` namespace.   

Marking approved so this can be merged once the small edits are made.

File: apf/apfH1Shapes.cc (hunk 1 above)
Comment (mortezah) [reply apf/apfH1Shapes.cc]: Fixed this.

File: apf/apfH1Shapes.cc (hunk 3 above)
Comment (mortezah) [reply apf/apfH1Shapes.cc]: Fixed this.

File: apf/apfH1Shapes.cc (hunk 4 above)
Comment (mortezah) [reply apf/apfH1Shapes.cc]: Fixed this.

File: apf/apfH1Shapes.cc (hunk 5 above)
Comment (mortezah) [reply apf/apfH1Shapes.cc]: Fixed this.

File: ma/maSolutionTransfer.cc (hunk 6 above)
Comment (mortezah) [reply ma/maSolutionTransfer.cc:135]: Fixed this.

File: apf/apfH1Shapes.cc (hunk 2 above)
Comment (mortezah) [reply apf/apfH1Shapes.cc]: Fixed this.

File: test/H1Shapes.cc (hunk 7 above)
Comment (mortezah) [reply test/H1Shapes.cc]: Fixed this.

File: test/testing.cmake (hunk 8 above)
Comment (mortezah) [reply test/testing.cmake]: These test only depend on the mesh. I have added the models to keep things uniform with the surrounding code.

**Comment by mortezah** [comment]
@cwsmith thanks for the reviews. I am going to merge this shortly if there are no more issues. Github has a few options

* merge commit
* squash and merge
* rebase and merge

Which one is the suggested method?

**Comment by cwsmith** [comment]
You're welcome.  

The answer depends on what history (sequence of commits) you want to exist in develop after the merge.

- `merge commit`will add the sequence of commits from the PR to develop + a 'merge' commit. 
- `squash and merge` will combine ('squash') all the commits from this PR into a new commit and add it and a 'merge' commit into develop.
- `rebase and merge` will squash the commits and add that commit onto develop

This appears like a decent, deeper, explanation of the differences:
https://rietta.com/blog/github-merge-types/

**Comment by mortezah** [comment]
@cwsmith thanks.

I will do the merge commit, since I will need to refer to some of the commits (for code style and development purposes) when I work with the students.

**Pull request merged**